api/internal/handler/upstream/upstream_test.go

```diff
@@ -19,285 +19,1719 @@ package upstream
+						},
+					},
+					Key: "server_addr",
+					Nodes: []interface{}{
+						map[string]interface{}{
+							"host":   "172.16.238.20",
+							"port":   float64(1981),
+							"weight": float64(1),
+						},
+					},
+				},
+			},
+			wantRet: &entity.Upstream{
+				BaseInfo: entity.BaseInfo{
+					ID: "u1",
+				},
+				UpstreamDef: entity.UpstreamDef{
+					Name: "upstream1",
+					Timeout: map[string]interface{}{
+						"connect": float64(20),
+						"send":    float64(20),
+						"read":    float64(20),
+					},
+					Checks: map[string]interface{}{
+						"active": map[string]interface{}{
+							"timeout":   float64(5),
+							"http_path": "/status",
+							"host":      "foo.com",
+							"healthy": map[string]interface{}{
+								"interval":  2,
+								"successes": 1,
+							},
+							"unhealthy": map[string]interface{}{
+								"interval":      1,
+								"http_failures": 2,
+							},
+							"req_headers": []interface{}{"User-Agent: curl/7.29.0"},
+						},
+						"passive": map[string]interface{}{
+							"healthy": map[string]interface{}{
+								"http_statuses": []interface{}{float64(200), float64(201)},
+								"successes":     float64(3),
+							},
+							"unhealthy": map[string]interface{}{
+								"http_statuses": []interface{}{float64(500)},
+								"http_failures": float64(3),
+								"tcp_failures":  float64(3),
+							},
+						},
+					},
+					Key: "server_addr_patch",
+					Nodes: []interface{}{
+						map[string]interface{}{
+							"host":   "172.16.238.20",
+							"port":   float64(1981),
+							"weight": float64(1),
+						},
+					},
+				},
+			},
+			getCalled: true,
+		},
+		{
+			caseDesc: "patch failed, path error",
+			giveInput: &PatchInput{
+				ID:      "u1",
+				SubPath: "error",
+				Body:    []byte("0"),
+			},
+			wantRet: handler.SpecCodeResponse(
+				errors.New("add operation does not apply: doc is missing path: \"error\": missing value")),
+			wantErr: errors.New("add operation does not apply: doc is missing path: \"error\": missing value"),
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.caseDesc, func(t *testing.T) {
+			getCalled := false
+
+			upstreamStore := &store.MockInterface{}
+			upstreamStore.On("Get", mock.Anything, mock.Anything).Return(existUpstream, nil)
+			upstreamStore.On("Update", mock.Anything, mock.Anything, mock.Anything).Run(func(args mock.Arguments) {
+				getCalled = true
+				input := args.Get(1).(*entity.Upstream)
+				createIfNotExist := args.Get(2).(bool)
+				assert.Equal(t, tc.wantInput, input)
+				assert.False(t, createIfNotExist)
+			}).Return(tc.giveRet, tc.giveErr)
+
+			h := Handler{upstreamStore: upstreamStore}
+			ctx := droplet.NewContext()
+			ctx.SetInput(tc.giveInput)
+			ret, err := h.Patch(ctx)
+			assert.Equal(t, tc.getCalled, getCalled)
+			assert.Equal(t, tc.wantRet, ret)
+			if tc.wantErr != nil && err != nil {
+				assert.Error(t, tc.wantErr.(error), err.Error())
+			} else {
+				assert.Equal(t, tc.wantErr, err)
+			}
```

Comment:
@Jaycean I'm a little confused. Could we use `assert.Equal(t, tc.wantErr, err)` here? Is this only for pretty logs for developers? Thanks.

Reply:
The patch method uses the package:`github.com/evanphx/json -Patch/V5`, when the error message is returned, the error is converted to match the error type, otherwise it cannot be matched.
Using wantret errors.New Create, geterr returns error.WithStack , error type with stack.
```
expected: *errors.errorString(&errors.errorString{s:"add operation does not apply: doc is missing path: \"error\": missing value"})
actual  : *errors.withStack(add operation does not apply: doc is missing path: "error": missing value)
```

Reply:
Yes.The same question I had meet.

Reply:
@Jaycean OK, gotcha. Thanks for the explanations.